### PR TITLE
Add code to extract .tar.gz files and link it to NPM and PyPI package managers

### PR DESCRIPTION
**Pull request by maxfisher-g**

### CODE REVIEW

File: internal/utils/archive_extract.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+// extractTar extracts the contents of the given stream of bytes of a tar archive.
+// If chdir is non-empty, the working directory will be temporarily changed to the
+// specified one before the extraction takes place, so that relative paths are resolved
+// relative to that directory. Otherwise, relative paths will be resolved relative to
+// the current working directory.
+func extractTar(tarStream io.Reader, chdir string) error {
+	var header *tar.Header
+	var err error
+
+	if chdir != "" {
```
Comment (oliverchang) [inline internal/utils/archive_extract.go:54]: Rather than changing dir and modifying global state, can we just support an `outputDir` directory that is written to without a chdir call? I.e. through computing the paths  through path.Join as we unpack.

Also, let's have callers be explicit and not support an empty `outputDir`.

File: internal/utils/archive_extract.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	for header, err = tarReader.Next(); err == nil; header, err = tarReader.Next() {
+		switch header.Typeflag {
+		case tar.TypeDir:
+			// make dir only readable by current user
+			if err = os.Mkdir(header.Name, 0700); err != nil {
+				return fmt.Errorf("mkdir failed: %w", err)
+			}
+		case tar.TypeReg:
+			var extractedFile *os.File
+			if extractedFile, err = os.Create(header.Name); err != nil {
```
Comment (oliverchang) [inline internal/utils/archive_extract.go:80]: We need to check for zip slip: https://security.snyk.io/research/zip-slip-vulnerability

One example of how this can be done is here: https://github.com/ossf/scorecard/blob/469216299ac3e161ebb995302ff8eb4ac43bb199/clients/githubrepo/tarball.go#L59

**Review by oliverchang — CHANGES_REQUESTED** [review]

File: internal/utils/archive_extract.go (hunk 1 above)
Comment (maxfisher-g) [reply internal/utils/archive_extract.go:54]: Yeah I agree changing global state isn't ideal, I just took inspiration from `tar -C`. I was also thinking that if the archive has absolute paths then these should be preserved, but actually on second thought they probably shouldn't in our case. So yes, given all this, let's do what you said.

File: internal/utils/archive_extract.go (hunk 2 above)
Comment (maxfisher-g) [reply internal/utils/archive_extract.go:80]: Good catch, will fix

File: internal/utils/archive_extract.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	tarReader := tar.NewReader(tarStream)
+
+	var header *tar.Header
+	var err error
+	for header, err = tarReader.Next(); err == nil; header, err = tarReader.Next() {
+		outputPath := filepath.Join(outputDir, header.Name)
+		// check for ZipSlip: https://snyk.io/research/zip-slip-vulnerability
+		if !strings.HasPrefix(outputPath, filepath.Clean(outputDir)) {
```
Comment (oliverchang) [inline internal/utils/archive_extract.go:63]: I'm not sure this works. 

If outputPath is "/outputPath/../foo", this passes the HasPrefix check right? 

I think this should be:

```golang
strings.HasPrefix(filepath.Clean(outputPath), filepath.Clean(outputDir)+string(os.PathSeparator)
```

`filePath.Clean` is needed to normalize both paths for comparison. e.g. "/outputPath/../foo" would turn into "/foo" before it's used for comparison. 

We should also add a os.Separator to the prefix to prevent "/outputPath/../outputPathFOO" from passing this check either.

Comment (oliverchang) [reply internal/utils/archive_extract.go:63]: Looking again at https://github.com/ossf/scorecard/blob/469216299ac3e161ebb995302ff8eb4ac43bb199/clients/githubrepo/tarball.go#LL49C23-L49C23, it looks like it's possible the scorecard implementation is wrong too.

Comment (maxfisher-g) [reply internal/utils/archive_extract.go:63]: Oh, not sure if you accounted for this but `filePath.Join()` also does a `Clean()` internally

Comment (maxfisher-g) [reply internal/utils/archive_extract.go]: Ok so turns out the first case was fine because of the fact that `filePath.Join()` does a `Clean()`, but adding `os.PathSeparator` was needed to avoid the second edge case. I've added tests for both cases

Comment (oliverchang) [reply internal/utils/archive_extract.go]: Ah yep, I misremembered the `filePath.Join()` behaviour. Sorry about that!

File: internal/utils/archive_extract_test.go [hunk 4]
```diff
@@ -74,65 +74,81 @@ func createTgzFile(name string, headers []*tar.Header) (path string, err error)
...
-				return
+			if n != size {
+				return fmt.Errorf("expected to write %d bytes but wrote %d bytes", size, n)
 			}
 		}
 	}
 
-	err = tarWriter.Close()
-	return
+	if err = tarWriter.Close(); err != nil {
```
Comment (oliverchang) [inline internal/utils/archive_extract_test.go:87]: Just `return tarWriter.Close()`

File: internal/utils/archive_extract_test.go [hunk 5]
```diff
@@ -74,65 +74,81 @@ func createTgzFile(name string, headers []*tar.Header) (path string, err error)
...
 	log.Initalize("")
-	err = ExtractTarGzFile(testFile, extractDir)
-	if err != nil {
+
+	if err = ExtractTarGzFile(archivePath, extractPath); err != nil {
 		return fmt.Errorf("extract failed: %v", err)
 	}
 
-	return runChecks(extractDir)
+	if err = runChecks(); err != nil {
```
Comment (oliverchang) [inline internal/utils/archive_extract_test.go:127]: just `return runChecks()` directly.

**Review by oliverchang — APPROVED** [review]

File: internal/utils/archive_extract_test.go (hunk 4 above)
Comment (maxfisher-g) [reply internal/utils/archive_extract_test.go:87]: Sure

File: internal/utils/archive_extract_test.go (hunk 5 above)
Comment (maxfisher-g) [reply internal/utils/archive_extract_test.go:127]: Can do

**Review by oliverchang — APPROVED** [review]
